agent/dns.go

```diff
@@ -1300,22 +1300,36 @@ func (d *DNSServer) handleRecurse(resp dns.ResponseWriter, req *dns.Msg) {
 	var r *dns.Msg
 	var rtt time.Duration
 	var err error
-	for _, recursor := range d.recursors {
+	for index, recursor := range d.recursors {
 		r, rtt, err = c.Exchange(req, recursor)
-		if err == nil || err == dns.ErrTruncated {
-			// Compress the response; we don't know if the incoming
-			// response was compressed or not, so by not compressing
-			// we might generate an invalid packet on the way out.
-			r.Compress = !d.disableCompression.Load().(bool)
-
-			// Forward the response
-			d.logger.Printf("[DEBUG] dns: recurse RTT for %v (%v)", q, rtt)
-			if err := resp.WriteMsg(r); err != nil {
-				d.logger.Printf("[WARN] dns: failed to respond: %v", err)
+		// Check if the response is valid and has the desired Response code
+		if r != nil && (r.Rcode != dns.RcodeSuccess && r.Rcode != dns.RcodeNameError) {
+			d.logger.Printf("[DEBUG] dns: recurse RTT for %v (%v) Recursor queried: %v Status returned: %v", q, rtt, recursor, dns.RcodeToString[r.Rcode])
+			// This is so that if the last known recursor also throws an error
```

Comment:
I dont think the body inside the `if {}` here is doing anything useful other than logging the recursor failure.

Condensed down this is the logic you have.

```
for ... {
   if ... {
      if last recursor {
         break
      }
      continue
   } else {
      if no error or truncated {
         write out the response
         return
      }
   }
}
```

The break if its the last recursor and continuing the loop will do the exact same thing which is to finish/exit the loop. This could be condensed into

```
for ... {
   if bad rcode {
      log
      continue
   }

   if no error or truncated {
      write out response
      return
   }
}
```

Reply:
Noted. I felt that the other if was totally unnecessary. Still wanted to get an opinion. Will add the test case and push it.